integrations/kube-agent-updater/pkg/cache/cache.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// StringAndErrorCache wraps a function returning (string, error) and caches the
+// result AND the error for a specific time. This cache is mainly used to ensure
+// external calls rates are reasonable. The cache is thread-safe.
+type StringAndErrorCache struct {
+	mutex         sync.Mutex
+	cachedVersion string
+	cachedError   error
+	validUntil    time.Time
+	cacheDuration time.Duration
+	getter        StringAndErrorGetter
+}
+
+// Get does a cache lookup and updates the cache in case of cache miss.
+func (c *StringAndErrorCache) Get(ctx context.Context) (string, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.validUntil.After(time.Now()) {
```

Comment:
Nit: If we used `clockwork` here, it would be easier to force expiry in tests

Reply:
Done (and I added basic cache tests)